### PR TITLE
fix(java): fix classloader get npe

### LINKED ISSUE #1763: NullPointer from ThreadPoolFury.serialize
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/fury/issues) and found no similar issues.

### Version

0.6.0

### Component(s)

Java

### Minimal reproduce step

I have not been able to produce a minimal code example. It is very difficult to debug, since I am just getting: `java.lang.NullPointerException: null` as the exception. This is occurring when trying to serialize values.

I am also seeing some errors when trying to deserialize values (far fewer then the serialize error):
```
java.lang.NullPointerException: Cannot invoke "String.length()" because "msg" is null
	at org.apache.fury.logging.FuryLogger.log(FuryLogger.java:145)
	at org.apache.fury.logging.FuryLogger.error(FuryLogger.java:133)
	at org.apache.fury.pool.FuryPooledObjectFactory.getPooledCache(FuryPooledObjectFactory.java:95)
	at org.apache.fury.pool.ThreadPoolFury.execute(ThreadPoolFury.java:77)
	at org.apache.fury.pool.ThreadPoolFury.deserialize(ThreadPoolFury.java:135)
```

### What did you expect to see?

No null pointers when serializing and deserializing data.

### What did you see instead?

j.l.NullPointerException: null

### Anything Else?

I am using Scala 2.13.12, my application runs in k8s (debian 11), with JDK 17. In production there are around 200 pods, and 20 cpus per pod.

I configure Fury in a scala Object like so:

```scala
object FuryUtils {
  lazy val fury = Fury
    .builder()
    .withLanguage(Language.JAVA)
    .withScalaOptimizationEnabled(true)
    .requireClassRegistration(false)
    .withRefTracking(false)
    .buildThreadSafeFuryPool(Runtime.getRuntime.availableProcessors() * 2, Runtime.getRuntime.availableProcessors() * 2)
}
```

and i serialize/deserialze my values like:
```scala
fury.serialize(value)
fury.deserialize(bytes).asInstanceOf[MyData]
```

I am calling this function ~20,000 times per second. And it is failing ~30 times per second.

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by chaokunyang**
## What does this PR do?

 fix classloader get npe

## Related issues

Closes #1763

## Does this PR introduce any user-facing change?

- [ ] Does this PR introduce any public API change?
- [ ] Does this PR introduce any binary protocol compatibility change?

### CODE REVIEW

File: java/fury-core/src/main/java/org/apache/fury/ThreadLocalFury.java [hunk 1]
```diff
@@ -257,6 +257,9 @@ public void setClassLoader(ClassLoader classLoader) {
 
   @Override
   public void setClassLoader(ClassLoader classLoader, StagingType stagingType) {
+    if (classLoader == null) {
```
Comment (LiangliangSui) [inline java/fury-core/src/main/java/org/apache/fury/ThreadLocalFury.java:260]: This should be redundant, there is already such a way in `org.apache.fury.util.LoaderBinding#setClassLoader`

File: java/fury-core/src/main/java/org/apache/fury/pool/ThreadPoolFury.java [hunk 2]
```diff
@@ -279,6 +279,9 @@ public void setClassLoader(ClassLoader classLoader) {
 
   @Override
   public void setClassLoader(ClassLoader classLoader, LoaderBinding.StagingType stagingType) {
+    if (classLoader == null) {
```
Comment (LiangliangSui) [inline java/fury-core/src/main/java/org/apache/fury/pool/ThreadPoolFury.java:282]: `org.apache.fury.pool.FuryPooledObjectFactory#setClassLoader`

File: java/fury-core/src/main/java/org/apache/fury/pool/FuryPooledObjectFactory.java [hunk 3]
```diff
@@ -50,7 +52,13 @@ public class FuryPooledObjectFactory {
 
   /** ThreadLocal: ClassLoader. */
   private final ThreadLocal<ClassLoader> classLoaderLocal =
-      ThreadLocal.withInitial(() -> Thread.currentThread().getContextClassLoader());
+      ThreadLocal.withInitial(() -> {
+        ClassLoader loader = Thread.currentThread().getContextClassLoader();
+        if (loader == null) {
+          loader = Fury.class.getClassLoader();
+        }
```
Comment (LiangliangSui) [inline java/fury-core/src/main/java/org/apache/fury/pool/FuryPooledObjectFactory.java:59]: Could we move this into `org.apache.fury.config.FuryBuilder#finish` ?

File: java/fury-core/src/main/java/org/apache/fury/ThreadLocalFury.java (hunk 1 above)
Comment (chaokunyang) [reply java/fury-core/src/main/java/org/apache/fury/ThreadLocalFury.java:260]: Good catch

**Comment by LiangliangSui** [comment]
Do we need to add the same logic in https://github.com/apache/fury/blob/main/java/fury-core/src/main/java/org/apache/fury/config/FuryBuilder.java#L323?

**Comment by chaokunyang** [comment]
> Do we need to add the same logic in https://github.com/apache/fury/blob/main/java/fury-core/src/main/java/org/apache/fury/config/FuryBuilder.java#L323?

Updated to:
```java
    if (classLoader == null) {
      classLoader = Thread.currentThread().getContextClassLoader();
      if (classLoader == null) {
        classLoader = Fury.class.getClassLoader();
      }
    }
```

**Review by LiangliangSui — APPROVED** [review]
LGTM

**Pull request merged**